api/src/main/java/io/cloudevents/CloudEvent.java

```diff
@@ -15,31 +15,62 @@
  */
 package io.cloudevents;
 
+import com.fasterxml.jackson.databind.JsonNode;
+
 import java.util.Map;
 import java.util.Optional;
 
 /**
  * An abstract event envelope
- * @param <A> The attributes type
- * @param <T> The 'data' type
  * @author fabiojose
+ * @author slinkydeveloper
  */
-public interface CloudEvent<A extends Attributes, T> {
+public interface CloudEvent {
 
     /**
      * The event context attributes
      */
-    A getAttributes();
+    Attributes getAttributes();
 
     /**
      * The event data
      */
-    Optional<T> getData();
+    Optional<String> getDataAsString() throws DataConversionException;
 
-    byte[] getDataBase64();
+    /**
+     * The event data
+     */
+    Optional<byte[]> getDataAsBytes() throws DataConversionException;
+
+    /**
+     * The event data
+     */
+    Optional<JsonNode> getDataAsJson() throws DataConversionException;
```

Comment:
since this is a shortcut to `event.getDataAsString().map(objectMapper::readTree)`, I'd advice against adding a strong dependency on Jackson to the public API.

It is fine to use Jackson internally, but, if we add it to the public API, we will stick to it for a looooooooong time

Reply:
It's not a shortuct for `getDataAsString()`, the reason to have it is that when the event is created from the json representation, there are events where the payload is already a json (like the third example here https://github.com/cloudevents/spec/blob/v1.0/json-format.md#32-examples). This allows to skip the marshalling to string and back. 

Be aware that in any case we'll have a dependency on jackson.

Reply:
dependency - yes, but Jackson type in the main public API...

Have you considered doing something like `<T> Optional<T> getDataAs(Class<T> clazz)`?

```java
event.getDataAs(JsonNode.class);
event.getDataAs(String.class);
event.getDataAs(byte[].class);
event.getDataAs(ByteBuffer.class);
```

would be highly optimizable on the implementation side while keeping the API JDK-only.

Reply:
To be honest i wish to avoid having a method like this because it could imply to have some encoding/decoding mechanism for the payload, similar to what we have in sdk-go (datacodec package).
TBH I'm prone to just have the payload as `[]byte`. It's easier both for the sdk and the users. It removes the optimization with json, but i think it's fine.

Reply:
yes now that i look at it i think it'a just better to have `byte[]` as data type